### PR TITLE
Add non-default noiseShape and seed support to dropout layer

### PR DESCRIPTION
**Pull request by syt123450**
This PR makes dropout layer support non-default `noiseShape` and `seed`, adds relative tests for these feature in `dropout` layer and backend's `dropout` op.

The `noiseShape` feature in `dropout` layer would benefit further development, for example, `SpatialDropout2D` layer, feature requested in [tensorflow/tfjs#1453](https://github.com/tensorflow/tfjs/issues/1453). Send this PR to ensure there won't be any broken change for further `noiseShape` related features.

**Note:**
This PR depends on [tensorflow/tfjs-core#1782](https://github.com/tensorflow/tfjs-core/pull/1782), which makes dropout op support non-default `noiseShape`.

---
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/556)

### CODE REVIEW

**Comment by caisq** [comment]
Gentle ping @syt123450

**Comment by caisq** [comment]
Gentle ping @syt123450 . Do you still plan to work on this? This PR is valuable. If so, please respond. If not, we'll need to close this PR as we are migrating tensorflow/tfjs-layers to tensorflow/tfjs to be unified with all other component repos of TensorFlow.js.

**Comment by syt123450** [comment]
Hi @caisq , thanks for the notification.

I have added tests to validate the usage of seed. As dropout op use randomUniform to generate random number, if seed if specified, we will get a specific value in training, I added checks for this number. 

I added tests both for backend dropout function and dropout layer. As I found that if I add value check in nested for loop, the codes will become extremely unreadable, so I added a separate test for a specified seed.

Sorry for late reply, I have really been stacking up with something recently, forgot the PRs. QAQ

**Comment by caisq** [comment]
@syt123450 Thanks for updating the PR. There seems to be a test failing:

Can you fix it?

> Step #1 - "test-browser": 	<Jasmine>
> Step #1 - "test-browser": 	_loop_5/<@src/layers/core_test.ts:65:23 <- src/layers/core_test.js:66:33
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/engine/topology.ts"]</Layer</Layer.prototype.apply@src/engine/topology.ts:976:12 <- src/engine/topology.js:719:16
> Step #1 - "test-browser": 	nameScope@src/common.ts:43:20 <- src/common.js:41:19
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/engine/topology.ts"]</Layer</Layer.prototype.apply/<@src/engine/topology.ts:1017:22 <- src/engine/topology.js:755:30
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call@src/layers/core.ts:89:12 <- src/layers/core.js:72:16
> Step #1 - "test-browser": 	tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:818:12
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8526:16
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.scopedRun@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8537:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy/<@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8527:22
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call/<@src/layers/core.ts:97:13 <- src/layers/core.js:78:30
> Step #1 - "test-browser": 	inTrainPhase@src/backend/tfjs_backend.ts:701:21 <- src/backend/tfjs_backend.js:652:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call/</output<@src/layers/core.ts:98:23 <- src/layers/core.js:78:66
> Step #1 - "test-browser": 	dropout@src/backend/tfjs_backend.ts:669:10 <- src/backend/tfjs_backend.js:619:12
> Step #1 - "test-browser": 	tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:818:12
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8526:16
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.scopedRun@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8537:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy/<@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8527:22
> Step #1 - "test-browser": 	dropout/<@src/backend/tfjs_backend.ts:669:21 <- src/backend/tfjs_backend.js:619:50
> Step #1 - "test-browser": 	f2@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:7396:26
> Step #1 - "test-browser": 	dropout_@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:7340:15
> Step #1 - "test-browser": 	Error: Non-default noise shape is not implemented yet: [2,3,1] in /tmp/karma-typescript-bundle-78k0JbNasOFmvm.js (line 7340)
> Step #1 - "test-browser": Firefox 66.0.0 (Mac OS X 10.13.0) Dropout Layer webgl2 {"WEBGL_VERSION":2,"WEBGL_CPU_FORWARD":false,"WEBGL_SIZE_UPLOAD_UNIFORM":0} tensor training=true, dropoutRate=0.5, noiseShape=[2,3,1] FAILED
> Step #1 - "test-browser": ...........
> Step #1 - "test-browser": 	<Jasmine>
> Step #1 - "test-browser": 	_loop_5/<@src/layers/core_test.ts:65:23 <- src/layers/core_test.js:66:33
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/engine/topology.ts"]</Layer</Layer.prototype.apply@src/engine/topology.ts:976:12 <- src/engine/topology.js:719:16
> Step #1 - "test-browser": 	nameScope@src/common.ts:43:20 <- src/common.js:41:19
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/engine/topology.ts"]</Layer</Layer.prototype.apply/<@src/engine/topology.ts:1017:22 <- src/engine/topology.js:755:30
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call@src/layers/core.ts:89:12 <- src/layers/core.js:72:16
> Step #1 - "test-browser": 	tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:818:12
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8526:16
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.scopedRun@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8537:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy/<@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8527:22
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call/<@src/layers/core.ts:97:13 <- src/layers/core.js:78:30
> Step #1 - "test-browser": 	inTrainPhase@src/backend/tfjs_backend.ts:701:21 <- src/backend/tfjs_backend.js:652:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/src/layers/core.ts"]</Dropout</Dropout.prototype.call/</output<@src/layers/core.ts:98:23 <- src/layers/core.js:78:66
> Step #1 - "test-browser": 	dropout@src/backend/tfjs_backend.ts:669:10 <- src/backend/tfjs_backend.js:619:12
> Step #1 - "test-browser": 	tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:818:12
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8526:16
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.scopedRun@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8537:23
> Step #1 - "test-browser": 	global.wrappers["/workspace/node_modules/@tensorflow/tfjs-core/dist/engine.js"]</Engine</Engine.prototype.tidy/<@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:8527:22
> Step #1 - "test-browser": 	dropout/<@src/backend/tfjs_backend.ts:669:21 <- src/backend/tfjs_backend.js:619:50
> Step #1 - "test-browser": 	f2@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:7396:26
> Step #1 - "test-browser": 	dropout_@/tmp/karma-typescript-bundle-78k0JbNasOFmvm.js:7340:15
> Step #1 - "test-browser": 	Error: Non-default noise shape is not implemented yet: [2,3,1] in /tmp/karma-typescript-bundle-78k0JbNasOFmvm.js (line 7340)
> Step #1 - "test-browser": Firefox 66.0.0 (Mac OS X 10.13.0) Dropout Layer cpu undefined tensor training=true, dropoutRate=0.5, noiseShape=[2,3,1] FAILED

**Comment by syt123450** [comment]
Hi @caisq , thanks for the review!

I think the test failed is caused by missing dependent PR. This PR depends on [tensorflow/tfjs-core#1782](https://github.com/tensorflow/tfjs-core/pull/1782), which hasn't been checked in.

I built a local tfjs-core and linked to tfjs-layers, and ran the tfjs-layers test suits locally, all tests passed, the screen shots shows my local test results:
Screen Shot 2019-08-08 at 12 47 47 PM

So I think we need to wait for the dependent PR, and wait for [tensorflow/tfjs-layers#576](https://github.com/tensorflow/tfjs-layers/pull/576) which is important for tfjs-layers to compatible with latest tfjs-core. If everything goes well, all tests would pass at that time.
